### ISSUE TITLE
Factory Method, Abstract Factory

### ISSUE THREAD

**Issue by MohamedAnwarAlhamed** [MohamedAnwarAlhamed/CSharp_Design_Patterns#1]
The Factory Design Pattern is one of the most widely used creational design patterns. It provides an interface for creating objects, but allows subclasses or other components to decide which specific class to instantiate. Instead of directly calling a constructor to create an object, the client refers to a factory method, which handles the logic of object creation. This promotes loose coupling and enhances flexibility.

Intent
-
The Factory Design Pattern is intended to:

Define an interface for creating an object, but let subclasses decide which class to instantiate.
Promote flexibility by decoupling the object creation process from the client.
Simplify object creation logic, especially when complex steps are involved.

Problem
-
When you create objects directly using constructors, the client needs to be aware of the exact class to instantiate, including its dependencies and configurations. This can make the client code tightly coupled to specific classes and difficult to maintain, particularly when object creation logic is complex or if the client has to choose from a variety of object types.

For example, consider an application that deals with different types of payment methods (Credit Card, PayPal, etc.). The client should not have to know which specific class (e.g., CreditCardPayment, PayPalPayment) to instantiate. It should simply request a payment method from a factory.

Solution
-
The Factory Design Pattern addresses this by introducing a factory method that the client uses to request objects. The factory method encapsulates the object creation logic, allowing the client to be unaware of the specific class or construction process behind the objects being created.

Structure
-
The Factory pattern has the following key participants:
- Product Interface: This is an abstract interface or class that defines the operations that the created objects (products) must implement.
- Concrete Products: These are the actual classes that implement the product interface. Each concrete product represents a different type of object that the factory can create.
- Creator (Factory): This defines the factory method, which returns an object of the Product type. The creator may be an abstract class or interface that allows subclasses to provide specific implementations of the factory method.
- Concrete Creators (Concrete Factories): These implement the factory method to return specific concrete product objects. The concrete creator decides which concrete product class to instantiate.

Define the Product Interface
```c#
// Product Interface
public interface ITransport
{
    void Deliver();
}

```

Define Concrete Products
```c#
// Concrete Product: Truck
public class Truck : ITransport
{
    public void Deliver()
    {
        Console.WriteLine("Delivery by land with a truck.");
    }
}

// Concrete Product: Ship
public class Ship : ITransport
{
    public void Deliver()
    {
        Console.WriteLine("Delivery by sea with a ship.");
    }
}

// Concrete Product: Plane
public class Plane : ITransport
{
    public void Deliver()
    {
        Console.WriteLine("Delivery by air with a plane.");
    }
}

```
Define the Creator (Factory)
```c#
// Creator (Factory)
public abstract class Logistics
{
    // Factory Method
    public abstract ITransport CreateTransport();

    // This method uses the factory method to deliver something
    public void PlanDelivery()
    {
        ITransport transport = CreateTransport();
        transport.Deliver();
    }
}

```

Define Concrete Creators (Concrete Factories)
```c#
// Concrete Creator: Road Logistics
public class RoadLogistics : Logistics
{
    public override ITransport CreateTransport()
    {
        return new Truck();
    }
}

// Concrete Creator: Sea Logistics
public class SeaLogistics : Logistics
{
    public override ITransport CreateTransport()
    {
        return new Ship();
    }
}

// Concrete Creator: Air Logistics
public class AirLogistics : Logistics
{
    public override ITransport CreateTransport()
    {
        return new Plane();
    }
}

```

Client Code
```c#
class Program
{
    static void Main(string[] args)
    {
        // Client wants Road logistics
        Logistics logistics = new RoadLogistics();
        logistics.PlanDelivery();

        // Client wants Sea logistics
        logistics = new SeaLogistics();
        logistics.PlanDelivery();

        // Client wants Air logistics
        logistics = new AirLogistics();
        logistics.PlanDelivery();
    }
}

```

Advantages of the Factory Design Pattern
-
- Loose Coupling: The client is decoupled from the object creation process. It only interacts with the factory, not the specific classes. This leads to better modularity and flexibility.
- Single Responsibility Principle: The factory encapsulates the object creation logic, allowing the client code to focus on using the objects rather than creating them.
- Centralized Creation Logic: The factory provides a single place to manage the object creation process. This makes the code more maintainable, as any changes to the creation logic can be made in one place.
- Open/Closed Principle: The factory can be extended to create new product types without modifying the existing client code. New ConcreteCreator and ConcreteProduct classes can be added without changing the existing system.
- Control Over Object Creation: The factory method can control how and when objects are created, adding extra logic such as caching, logging, or configuring objects based on the context.

Disadvantages of the Factory Design Pattern
-
- Complexity: The pattern introduces additional classes and complexity into the system, especially when dealing with simple applications where object creation is straightforward.
- Overhead: If not used judiciously, factories can add unnecessary abstraction, making the code harder to understand and maintain.

When to Use the Factory Design Pattern
-
- When you have a group of objects with similar properties: The factory pattern is useful when multiple classes implement the same interface but differ in their implementation details.
- When the creation logic is complex: If the process of creating an object involves many steps or complex logic, the factory pattern can encapsulate this complexity.
- When object creation needs to be decoupled from usage: If the client code should not be concerned with the specific class to instantiate, the factory pattern is ideal.
- When you expect to add new products later: If the system is likely to expand with new types of products, using a factory makes it easy to add new classes without modifying the existing client code.

Real-World Examples of the Factory Pattern
-
- Java's Calendar class: In Java, the Calendar class uses a factory method getInstance() to create different types of calendar systems (e.g., GregorianCalendar).
- Document Creation: Many applications (e.g., Microsoft Word) use the factory pattern to create different types of documents (text files, spreadsheets, presentations). The client simply requests a document from the factory without needing to know the specific class that implements it.
- Database Connections: A factory might be used to create connections to different databases (e.g., MySQL, PostgreSQL) based on a configuration, without requiring the client to handle the details of each type of connection.
- UI Components: UI toolkits often use factories to create components like buttons, windows, and menus. The specific look and feel of the components might vary based on the platform, but the client code interacts with a factory to request the components.

In summary, the Factory Design Pattern is a powerful tool for managing object creation, promoting flexibility, and decoupling the client code from the specific classes it interacts with.
